accepted/cranelift-roadmap-2022.md

```diff
@@ -0,0 +1,895 @@
+improve even more.
+   
+A particularly valuable concrete milestone would be to stabilize our
+SIMD support and turn it on by default. Beyond that, the sky is the
+limit: any sort of special-case optimization that would have been more
+difficult to express in the handwritten pattern-matching code can now
+be added in a few lines in the DSL.
+
+Two likely useful approaches will be to (i) mine LLVM for existing,
+well-tested and well-thought-out sets of optimized lowerings, and (ii)
+leverage superoptimization tools such as
+[Souper](https://github.com/google/souper) to generate lowerings.
+   
+## Improved Optimizer (Compiler Middle-End)
+
+Second, Cranelift needs to improve its "middle end", or
+machine-independent optimization stage. Most of our effort in the past
+several years has been in improving the backends, but our "standard
+suite of optimizations" -- GVN, DCE, LICM, constant
+propagation/folding, and the like -- have not changed in that
+time. This has served us perfectly fine for a while, but both new
+workload characteristics, and increasing focus on the parts of the
+pipeline that we haven't recently optimized (given all the work on the
+backends), indicate that perhaps this should change.
+
+In particular, the WebAssembly MVP design, which drove Cranelift's
+design, implies some interesting code characteristics. Because a Wasm
+module is typically (not always!) produced by an optimizing compiler
+such as LLVM, heavyweight optimizations are *usually* not very
+worthwhile. For example, trivial helper functions will have already
+been inlined before the Wasm module was produced, so inlining is less
+important for a Wasm engine to implement. Similar arguments can be
+made for heavyweight strength-reduction and algebraic-identity
+optimizations, constant propagation, LICM, and the like. Translation
+from Wasm to CLIF *does* introduce some redundancies, for example in
+heap address computations, but Cranelift's simple optimizations can
+handle many of these issues.
+
+However, multi-module WebAssembly workloads will become more common as
+[Module Linking](https://github.com/WebAssembly/module-linking) is
+used by real workloads and as the [Interface
+Types](https://github.com/WebAssembly/interface-types) idea becomes
+real. This changes everything: when we have a late-binding
+environment, in which two Wasm modules that have never met each other
+before are linked together due to user configuration, all of the usual
+compiler optimizations may apply to this new combination of code.
+
+### Inlining
+
+An inliner is a necessary prerequisite to any optimizations across
+modules; otherwise, we can't take advantage of any knowledge that
+derives from the combination of modules (unless we do interprocedural
+analysis, which we are unlikely to ever do).
+   
+So, we should develop a simple inliner, with the unique design
+constraint that it only needs to work well in cross-module
```

Comment:
There is also another design constraint: It shouldn't break parallel compilation to ensure that we don't drastically regress compilation times. Function compilation is currently entirely parallelizable as everything is local. Inlining however is an inherently non-local optimization.

Reply:
That's a good point, thanks! I added a paragraph that describes the tradeoff. The three design points I can think of right away are

- Some sort of snapshotting, so that each function can be mutated separately while also accessing the function-to-be-inlined in a read-only way; or
- Some kind of partially-parallel design, where we "stratify" the callgraph into a DAG of SCCs then scan a parallel frontier of inlining up the DAG, from leafs to roots, processing serially within each SCC (this is maximally parallel while still allowing the inliner to see already-inlined versions of functions it might in turn inline); or
- Make inlining a special, optional, more costly pass that just operates serially, and takes a mutable borrow of the whole module or modules (collection of functions). This might be the best way to start.